crates/qasm3/src/build.rs

```diff
@@ -226,42 +226,37 @@ impl BuilderState {
         self.qc.append(py, instruction).map(|_| ())
     }
 
-    fn define_gate(
-        &mut self,
-        _py: Python,
-        ast_symbols: &SymbolTable,
-        decl: &asg::GateDeclaration,
-    ) -> PyResult<()> {
-        let name_id = decl
-            .name()
-            .as_ref()
-            .map_err(|err| QASM3ImporterError::new_err(format!("internal error: {:?}", err)))?;
-        let name_symbol = &ast_symbols[name_id];
-        let pygate = self.pygates.get(name_symbol.name()).ok_or_else(|| {
-            QASM3ImporterError::new_err(format!(
-                "can't handle non-built-in gate: '{}'",
-                name_symbol.name()
-            ))
-        })?;
-        let defined_num_params = decl.params().as_ref().map_or(0, Vec::len);
-        let defined_num_qubits = decl.qubits().len();
-        if pygate.num_params() != defined_num_params {
-            return Err(QASM3ImporterError::new_err(format!(
-                "given constructor for '{}' expects {} parameters, but is defined as taking {}",
-                pygate.name(),
-                pygate.num_params(),
-                defined_num_params,
-            )));
-        }
-        if pygate.num_qubits() != defined_num_qubits {
-            return Err(QASM3ImporterError::new_err(format!(
-                "given constructor for '{}' expects {} qubits, but is defined as taking {}",
-                pygate.name(),
-                pygate.num_qubits(),
-                defined_num_qubits,
-            )));
+    // Map gates in the symbol table to Qiskit gates in the standard library.
+    // Encountering any gates not in the standard library results in raising an exception.
+    // Gates mapped via CustomGates will not raise an exception.
+    fn map_gate_ids(&mut self, _py: Python, ast_symbols: &SymbolTable) -> PyResult<()> {
+        for (name, name_id, defined_num_params, defined_num_qubits) in ast_symbols.gates().iter() {
```

Comment:
Assuming `SymbolTable::gates` produces some sort of `&[T]`, the `.iter()` probably isn't necessary because `into_iter()` (which `for` calls implicitly) on `&[T]` consumes the reference but not the `T`, so there's no borrow-check problem.

Reply:
[Here is `SymbolTable::gates`](https://github.com/Qiskit/openqasm3_parser/blob/92dead66a0f27d5a12ec301d10293f6e453b0034/crates/oq3_semantics/src/symbols.rs#L262-L276) : 
```rust
    /// Return a Vec of information about all gate declarations. Each element
    /// is a tuple of (gate name, symbol id, num classical params, num quantum params).
    pub fn gates(&self) -> Vec<(&str, SymbolId, usize, usize)> {
```
It ends with `collect()` because I was unable to quickly come up with a trait bound for the return type.... or maybe I never even tried for this one. But I think it is more flexible to return the iterator, right?

Reply:
I removed `.iter()`. 
Fixed in d7a5e61

Reply:
I also removed the final `collect` in `SymbolTable::gates`. The return type is no longer `Vec`, but is now `impl Iterator<Item = (&str, SymbolId, usize, usize)>`. But this required no further edits to the present PR. I mean d7a5e61 was enough. This change is present in oq3_semantics 0.6.0. The present PR depends on this version.

Reply:
Ah, that's good to both removing the `iter` and making the function return some unnamed `impl Iterator` thanks.